File: day01/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
)

const (
	puzzleInput = "input.txt"
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatal(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatal(err)
		}
	}()

	prev := 0
	prev1 := 0
	prevsum := 0
	count := -1
	count2 := -3
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		num, err := strconv.Atoi(scanner.Text())
		if err != nil {
			log.Fatal(err)
		}
		if num > prev {
			count++
		}
		k := num + prev + prev1
		if k > prevsum {
			count2++
		}
		prevsum = k
		prev1 = prev
		prev = num
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	fmt.Println("Part 1:", count)
	fmt.Println("Part 2:", count2)
}

File: day01/src/main.rs
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let nums = {
        let mut nums = Vec::new();
        for line in reader.lines() {
            nums.push(line?.parse::<i32>()?);
        }
        nums
    };

    let mut count = 0;
    for [prev, next] in nums.windows(2).flat_map(<&[i32; 2]>::try_from) {
        if next > prev {
            count += 1;
        }
    }
    let mut count2 = 0;
    for [prev, next] in nums
        .windows(3)
        .flat_map(<&[i32; 3]>::try_from)
        .map(|[one, two, three]| one + two + three)
        .collect::<Vec<i32>>()
        .windows(2)
        .flat_map(<&[i32; 2]>::try_from)
    {
        if next > prev {
            count2 += 1;
        }
    }
    println!("Part 1: {}", count);
    println!("Part 2: {}", count2);
    Ok(())
}
